---
title: "Working with Sample Data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{sample_data}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Background

In `qSIP2`, "sample data" refers to any high level metadata associated with either an experiment or the individual fractions. This vignette will show available tools to format and validate your sample data for the `qsip_sample_data` class in the `qSIP2` package.

```{r setup, message=FALSE}
library(dplyr)
library(qSIP2)
```

## What is a sample?

The word **sample** typically refers to the biological or environmental entity the DNA was isolated from as well as the single sequencing run tied to that **sample**. In qSIP, however, because there are multiple sequencing runs per biological subject, the term **sample** has historically been reserved for sequencing of each fraction. In practice, this means you will have many **samples** for each single biological subject.

Traditionally, the original biological subject in qSIP has been simply called a **replicate** or "rep". In the proposed MISIP^[https://www.biorxiv.org/content/10.1101/2023.07.13.548835v1] standards, the concept of a **rep** has been replaced with `source_mat_id`.

For example, if you were identifying active community members from a pot of soil in triplicate, you may have 3 soil pot **reps** that were given <sup>16</sup>O (maybe named "A_16O", "B_16O" and "C_16O"), and then 3 **reps** with <sup>18</sup>O (maybe named "A_18O", "B_18O" and "C_18O"). In this experiment, you would have 6 totals **reps**. If each **rep** was separated into 20 fractions and sequenced, then you would have 120 (6x20) **samples**.

If working with amplicon data, each fastq file would be a **sample**, or each column in an ASV/OTU table (when taxa are the rows).

# Creating a standardized sample data tibble

To standardize the qSIP workflow, column names should adhere as closely to MISIP standards where possible. Absolute requirements are slightly different than the official MISIP standards, and for the `qSIP2` package would be the `source_mat_id`, `isotope`, `gradient_position`, `gradient_pos_density` and `gradient_pos_rel_amt` fields.

## isotope and isotopolog_label fields

In traditional qSIP the `isotope` field has been populated with either the light (e.g. 16O) or heavy (e.g. 18O) isotope depending on the substrate used in that rep or `source_mat_id`. In MISIP standards, only the heavy isotope is listed under the `isotope` field, and then a secondary field `isotopolog_label` is used to designate whether the replicate used a substrate with "natural abundance" (i.e. "light") or "isotopically labeled" (i.e. "heavy") isotopes.

In the `qSIP2` package, either method can be used. If the `isotopolog_label` is missing from your dataset then it will assume both the light and heavy isotopes are present in the `isotope` field. But, if you do have an `isotopolog_label` field, then only the heavy isotope designation is allowed in the `isotope` field and the dataframe will not pass validation checks if there are light isotopes listed there. 

Conversion between these two objects can be done with the `add_isotopolog_label()` or `remove_isotopolog_label()` functions.

```{r eval=F}
# original data has a mixture of isotopes
head(sample_data_nonMISIP)

sample_data_nonMISIP %>%
  count(isotope)

# new data has only one isotope and a mixture of isotopolog_label
df_with_labels <- add_isotopolog_label(sample_data_nonMISIP,
  isotope = "isotope"
)

df_with_labels %>%
  count(isotope, isotopolog_label)

# after converting back the isotope column is a mixture again
remove_isotopolog_label(df_with_labels) %>%
  count(isotope)
```

## gradient_pos_rel_amt field

A requirement for qSIP is the `gradient_pos_rel_amt` field, which gives the percent amount that a fraction has of the whole. The preferred method is given in qPCR copy numbers, but DNA concentrations can be used as well. 

For example, if there are 100,000 total 16S copies in a replicate as determined by qPCR, and 15,000 copies in fraction 7, then the `gradient_pos_rel_amt` value for fraction 7 would be 0.15 (15,000 / 100,000). Similarly, if you had 25ng total DNA used for density separation, and fraction 7 had 3.75 ng DNA recovered, then `gradient_pos_rel_amt` would also be 0.15 (3.75 / 25). 

Ideally, all of the `gradient_pos_rel_amt` for a given `source_mat_id` should add up to 1, but there are situations where it might be less than 1. For example, if you removed some fractions because they didn't sequence well or there was some other reason to remove a fraction. If you have 16S or DNA concentrations for these removed samples they would be subtracted from the total. So, for our example above, if fraction 7 needed to be removed, then the total for all fractions of that `source_mat_id` would only be 0.85. Another situation would be that although you added 25ng to a centrifugation, you only recovered 20ng and the rest was lost[^angel_share]. The total fractions within a `source_mat_id` within should never be greater than 1.

[^angel_share]: [The angel's share and devil's cut](https://www.visitlex.com/guides/post/common-geeky-bourbon-terms-explained/#:~:text=Angel's%20Share)

If your sample data tibble does not have the `gradient_pos_rel_amt` there is a function that can add it for you. 

The `add_gradient_pos_rel_amt()` can create this column from either a column of qPCR totals, or DNA concentrations. Because this function can not know if there are missing fractions, the totals per `source_mat_id` will be equal to 1. If you do have fractions that you want removed, keep them in the dataframe for this `add_gradient_pos_rel_amt()` and then remove them after.

```{r, error = TRUE, eval=F}
# gradient_pos_rel_amt to a dataframe without this column
add_gradient_pos_rel_amt(sample_data_nonMISIP,
  source_mat_id = "rep", # the source_mat_id or per-tube grouping
  amt = "dna_conc" # the column with total amounts
) %>%
  head()

# adding to a dataframe with an existing gradient_pos_rel_amt will give an error
add_gradient_pos_rel_amt(sample_data_MISIP,
  amt = "gradient_pos_dna"
)

# adding overwrite = T removes the error
add_gradient_pos_rel_amt(sample_data_MISIP,
  amt = "gradient_pos_dna",
  overwrite = T
) %>%
  head()
```







# Make a qSIP sample data object

The `qsip_sample_data` constructor builds a sample data object from a dataframe. The `qSIP2` package comes with a basic example tibble called `sample_data_MISIP`, and some of the columns are already formatted with proper MISIP headers.

```{r eval=F}
# qdf = qsip_sample_data(sample_data_MISIP)
```

And that's that, for now.

## Isotope Values


