<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Growth Workflow • qSIP2</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Growth Workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">qSIP2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.20.5.9004</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/qSIP_workflow.html">Standard qSIP EAF Workflow</a></li>
    <li><a class="dropdown-item" href="../articles/filtering.html">Filtering Features</a></li>
    <li><a class="dropdown-item" href="../articles/resampling.html">Resampling</a></li>
    <li><a class="dropdown-item" href="../articles/EAF.html">EAF Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/growth.html">Growth Workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jeffkimbrel/qSIP2/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Growth Workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jeffkimbrel/qSIP2/blob/master/vignettes/growth.Rmd" class="external-link"><code>vignettes/growth.Rmd</code></a></small>
      <div class="d-none name"><code>growth.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jeffkimbrel/qSIP2" class="external-link">qSIP2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"qSIP2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] '0.20.5.9004'</span></span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>Using qSIP2, we can estimate the growth rate of individual features
(i.e. bacteria) in a microbial community by fitting a growth model to
the abundance of a labeled taxon over time. Assumptions for growth
include</p>
<ol style="list-style-type: decimal">
<li>there is no isotopic labeling at time zero</li>
<li>the pool of unlabeled features will not go up over time</li>
<li>bacteria that incorporate the isotope are 100% labeled</li>
</ol>
<p>Using the calculated EAF values from the standard workflow, we
therefore can say an EAF of 0.5 means that 50% of the bacteria are
labeled and the result of growth since time zero (i.e. “birth” or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>b</mi><annotation encoding="application/x-tex">b</annotation></semantics></math>).
Further, using quantitative abundance values in both time zero and time
point samples, we can estimate the death rate
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>d</mi><annotation encoding="application/x-tex">d</annotation></semantics></math>)
of individual features by calculating the decrease in unlabeled
features. Together, we get at the growth rate using the equation
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>r</mi><mi>i</mi></msub><mo>=</mo><msub><mi>b</mi><mi>i</mi></msub><mo>+</mo><msub><mi>d</mi><mi>i</mi></msub></mrow><annotation encoding="application/x-tex">r_i = b_i + d_i</annotation></semantics></math>
for each feature
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://esajournals.onlinelibrary.wiley.com/doi/full/10.1002/ecs2.2090" class="external-link uri"&gt;https://esajournals.onlinelibrary.wiley.com/doi/full/10.1002/ecs2.2090&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a>. This is
one of the main advantages of qSIP where if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>b</mi><annotation encoding="application/x-tex">b</annotation></semantics></math>
equals
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>d</mi><annotation encoding="application/x-tex">d</annotation></semantics></math>
then traditional community analysis would detect no change in the
community, whereas qSIP would detect growth and death of individual
features.</p>
<p>Growth analysis starts with calculations of EAF values, and therefore
requires the standard workflow (<code><a href="../articles/qSIP_workflow.html">vignette("qSIP_workflow")</a></code>)
to be run first with only slight modifications.</p>
<p>For growth, three additional arguments are required for the
<code>qsip_source_data</code> creation.</p>
<ol style="list-style-type: decimal">
<li>
<code>timepoint</code> - a numerical value for the timepoint of the
source material. There is often a <code>0</code> timepoint, but these
can be any values and the growth rate will be calculated as the
difference between time points. Further, these can be any units
(e.g. days, hours, etc.), and the interpretation of the growth rate will
depend on the units (e.g. “per day” or “per hour”).</li>
<li>
<code>total_abundance</code> - a numerical value for the total
abundance of the source material. Ideally, this is a copy number from
qPCR using the same primers as the sequencing. Further, it should be
standardized to some unit of starting material (e.g. copies per gram of
soil). If it isn’t, then the next <code>volume</code> argument is
important.</li>
<li>
<code>volume</code> - a numerical value for the volume of the source
material DNA that the copy number was derived from. Typically the volume
is the same for all source material DNA extractions, but if your
starting volume for qPCR was different then this parameter is
important.</li>
</ol>
</div>
<div class="section level2">
<h2 id="growth-object">Growth Object<a class="anchor" aria-label="anchor" href="#growth-object"></a>
</h2>
<p>An example growth object is provided with the <code>qSIP2</code>
package called <code>example_qsip_growth_object</code>. We can check
which columns contain the three additional arguments for growth, and
pull out a table with the relevant columns.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_dataframe.html">get_dataframe</a></span><span class="op">(</span><span class="va">example_qsip_growth_object</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">source_mat_id</span>, <span class="va">isotope</span>, <span class="va">timepoint</span>, <span class="va">total_abundance</span>, <span class="va">volume</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">timepoint</span>, <span class="va">isotope</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">source_mat_id</th>
<th align="left">isotope</th>
<th align="right">timepoint</th>
<th align="right">total_abundance</th>
<th align="right">volume</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">source_1</td>
<td align="left">Time0</td>
<td align="right">0</td>
<td align="right">20934337125</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">source_10</td>
<td align="left">Time0</td>
<td align="right">0</td>
<td align="right">56376407410</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">source_13</td>
<td align="left">Time0</td>
<td align="right">0</td>
<td align="right">7952816086</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">source_4</td>
<td align="left">Time0</td>
<td align="right">0</td>
<td align="right">38061061332</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">source_7</td>
<td align="left">Time0</td>
<td align="right">0</td>
<td align="right">28775383886</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">source_11</td>
<td align="left">16O</td>
<td align="right">10</td>
<td align="right">5053795437</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">source_14</td>
<td align="left">16O</td>
<td align="right">10</td>
<td align="right">219349821</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">source_2</td>
<td align="left">16O</td>
<td align="right">10</td>
<td align="right">5006451196</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">source_5</td>
<td align="left">16O</td>
<td align="right">10</td>
<td align="right">5440927504</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">source_8</td>
<td align="left">16O</td>
<td align="right">10</td>
<td align="right">3000381981</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">source_12</td>
<td align="left">18O</td>
<td align="right">10</td>
<td align="right">5043787157</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">source_15</td>
<td align="left">18O</td>
<td align="right">10</td>
<td align="right">200708494</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">source_3</td>
<td align="left">18O</td>
<td align="right">10</td>
<td align="right">5524820407</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">source_6</td>
<td align="left">18O</td>
<td align="right">10</td>
<td align="right">5242785770</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">source_9</td>
<td align="left">18O</td>
<td align="right">10</td>
<td align="right">3702908766</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<p>From this table, we can notice a few things. First, there are 15
total samples - 5 with timepoint 0, and 5 each with 16O or 18O isotopes.
Second, some sources do not have a standard <code>isotope</code>
designation, but instead say “Time0”. This is a special allowed
<code>isotope</code> type flagging these sources as unfractionated, and
therefor no EAF value will be calculated for them. Third, the
<code>volume</code> column is the same for all samples which indicates
that the <code>total_abundance</code> is already standardized to the
same volume. Indeed if we look at the column that
<code>total_abundance</code> was derived from we can tell from the name
that it is a copy number to a standardize amount of soil (16S copies per
gram of soil).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_qsip_growth_object</span><span class="op">@</span><span class="va">source_data</span><span class="op">@</span><span class="va">total_abundance</span></span>
<span><span class="co">#&gt; [1] "qPCR.16S.copies.g.soil"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="eaf-workflow">EAF Workflow<a class="anchor" aria-label="anchor" href="#eaf-workflow"></a>
</h2>
<p>As mentioned above, the growth workflow requires the EAF values to be
calculated first. Note, we are running with
<code>allow_failures = TRUE</code>, but still with a minimum of 4
labeled and 4 unlabeled fractions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_feature_filter.html">run_feature_filter</a></span><span class="op">(</span><span class="va">example_qsip_growth_object</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"Day 10"</span>,</span>
<span>  unlabeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"source_11"</span>, <span class="st">"source_14"</span>, <span class="st">"source_2"</span>, <span class="st">"source_5"</span>, <span class="st">"source_8"</span><span class="op">)</span>,</span>
<span>  labeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"source_12"</span>, <span class="st">"source_15"</span>, <span class="st">"source_3"</span>, <span class="st">"source_6"</span>, <span class="st">"source_9"</span><span class="op">)</span>,</span>
<span>  min_labeled_fractions <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  min_unlabeled_fractions <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/run_resampling.html">run_resampling</a></span><span class="op">(</span></span>
<span>    resamples <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    with_seed <span class="op">=</span> <span class="fl">1332</span>,</span>
<span>    allow_failures <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    progress <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/run_EAF_calculations.html">run_EAF_calculations</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; There are initially 364 unique feature_ids</span></span>
<span><span class="co">#&gt; 364 of these have abundance in at least one fraction of one source_mat_id</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; Filtering feature_ids by fraction...</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">15</span> unlabeled and <span style="color: #BB0000;">11</span> labeled feature_ids were found in <span style="color: #BB0000;">zero fractions</span> in at least one source_mat_id</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">70</span> unlabeled and <span style="color: #BB0000;">47</span> labeled feature_ids were found in <span style="color: #BB0000;">too few fractions</span> in at least one source_mat_id</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">364</span> unlabeled and <span style="color: #00BB00;">364</span> labeled feature_ids <span style="color: #00BB00;">passed</span> the fraction filter</span></span>
<span><span class="co">#&gt; In total, <span style="color: #00BB00;">364</span> unique feature_ids <span style="color: #00BB00;">passed</span> the fraction filtering requirements...</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; Filtering feature_ids by source...</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">6</span> unlabeled and <span style="color: #BB0000;">5</span> labeled feature_ids failed the source filter because they were found in <span style="color: #BB0000;">too few sources</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">358</span> unlabeled and <span style="color: #00BB00;">359</span> labeled feature_ids <span style="color: #00BB00;">passed</span> the source filter</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; In total, <span style="color: #00BB00;">358</span> unique feature_ids <span style="color: #00BB00;">passed</span> all fraction and source filtering requirements</span></span>
<span><span class="co">#&gt; Warning: 8 unlabeled and 2 labeled feature_ids had resampling failures. Run</span></span>
<span><span class="co">#&gt; `get_resample_counts()` or `plot_successful_resamples()` on your &lt;qsip_data&gt;</span></span>
<span><span class="co">#&gt; object to inspect.</span></span></code></pre></div>
<p>Overall, most features had robust resampling results, with only a few
having less than 99% success in the labeled sources.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_resample_counts.html">get_resample_counts</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">labeled_resamples</span> <span class="op">&lt;</span> <span class="fl">1000</span> <span class="op">|</span> <span class="va">unlabeled_resamples</span> <span class="op">&lt;</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 8 × 3</span></span></span>
<span><span class="co">#&gt;   feature_id labeled_resamples unlabeled_resamples</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> taxon_113               <span style="text-decoration: underline;">1</span>000                 999</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> taxon_180               <span style="text-decoration: underline;">1</span>000                 999</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> taxon_234                996                 990</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> taxon_250                993                 990</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> taxon_278               <span style="text-decoration: underline;">1</span>000                 999</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> taxon_292               <span style="text-decoration: underline;">1</span>000                 999</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> taxon_327               <span style="text-decoration: underline;">1</span>000                 999</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> taxon_341               <span style="text-decoration: underline;">1</span>000                 999</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_EAF_values.html">plot_EAF_values</a></span><span class="op">(</span><span class="va">q</span>,</span>
<span>  confidence <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  error <span class="op">=</span> <span class="st">"ribbon"</span>,</span>
<span>  success_ratio <span class="op">=</span> <span class="fl">0.9</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span></code></pre></div>
<p><img src="growth_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="growth-workflow">Growth Workflow<a class="anchor" aria-label="anchor" href="#growth-workflow"></a>
</h2>
<div class="section level3">
<h3 id="time-zero-total-abundances">Time zero total abundances<a class="anchor" aria-label="anchor" href="#time-zero-total-abundances"></a>
</h3>
<p>In addition to the EAF values stored in the <code>qsip_data</code>
object, we also need a table with the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mrow><mi>T</mi><mi>O</mi><mi>T</mi><mi>A</mi><mi>L</mi><mi>i</mi><mn>0</mn></mrow></msub><annotation encoding="application/x-tex">N_{TOTALi0}</annotation></semantics></math>
values for each feature
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
at timepoint
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>,
in this case time 0. This value is the <em>total</em> abundance of each
feature and is the sum of both the labeled and unlabeled features
(equation 2 from Koch, 2018<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://esajournals.onlinelibrary.wiley.com/doi/full/10.1002/ecs2.2090" class="external-link uri"&gt;https://esajournals.onlinelibrary.wiley.com/doi/full/10.1002/ecs2.2090&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>). Note you don’t have to always compare
against time zero. If you have a 7-day and 14-day timepoint you can set
day 7 as the initial timepoint here.</p>
<p>This table is created with the <code><a href="../reference/get_N_total_it.html">get_N_total_it()</a></code> function
where you pass the original <code>qsip_data</code> object and the
timepoint of interest.</p>
<div class="alert alert-primary">
<p><code><a href="../reference/get_N_total_it.html">get_N_total_it()</a></code> should be run on the initial
<code>qsip_data</code> object before any filtering or resampling has
been done. This is because the unfractionated time zero sources will not
be present in the filtered data.</p>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N_total_i0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_N_total_it.html">get_N_total_it</a></span><span class="op">(</span><span class="va">example_qsip_growth_object</span>, t <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: 1 feature_ids have zero abundance at time 0:</span></span>
<span><span class="co">#&gt; Warning: taxon_194</span></span></code></pre></div>
<p>Note we get a warning here that taxon_194 has zero abundance at
<code>t = 0</code>. Therefore, this feature cannot have a growth rate
calculated because any change in abundance would be considered infinite
growth.</p>
<table class="table">
<caption>First few rows of <code>N_total_i0</code>
</caption>
<thead><tr class="header">
<th align="left">feature_id</th>
<th align="right">N_total_i0</th>
<th align="right">timepoint1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">taxon_1</td>
<td align="right">1595472105</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">taxon_2</td>
<td align="right">64576684</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">taxon_3</td>
<td align="right">4488930</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">taxon_4</td>
<td align="right">2494463</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">taxon_5</td>
<td align="right">9849881</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">taxon_6</td>
<td align="right">697760597</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="growth-rate-calculations">Growth rate calculations<a class="anchor" aria-label="anchor" href="#growth-rate-calculations"></a>
</h3>
<p>Using the abundance values stored in <code>N_total_i0</code> and the
resampled EAF values stored in <code>q</code>, we can calculate the
growth rate for each feature. This is done with the
<code><a href="../reference/run_growth_calculations.html">run_growth_calculations()</a></code> function where you pass the
<code>qsip_data</code> object, the <code>N_total_i0</code> table, and
the growth model to use. The growth model can be either “exponential” or
“linear”.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_growth_calculations.html">run_growth_calculations</a></span><span class="op">(</span><span class="va">q</span>,</span>
<span>                               N_total_it <span class="op">=</span> <span class="va">example_qsip_growth_t0</span>,</span>
<span>                               growth_model <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: 31862 resamplings have a negative EAF value or calculated labeled copy</span></span>
<span><span class="co">#&gt; numbers less than 0. These values have been filtered out and added to</span></span>
<span><span class="co">#&gt; @growth$negative_labeled</span></span></code></pre></div>
<p>Note the warning message, which we will return to in a minute.</p>
</div>
<div class="section level3">
<h3 id="growth-calculation-results">Growth calculation results<a class="anchor" aria-label="anchor" href="#growth-calculation-results"></a>
</h3>
<p>We can get a dataframe of the growth calculations with the
<code><a href="../reference/get_growth_data.html">get_growth_data()</a></code> function. Here, we will also filter to
just the data for the first resample.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_growth_data.html">get_growth_data</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">resample</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>The first few rows of <code>get_growth_data(q)</code>
</caption>
<colgroup>
<col width="7%">
<col width="7%">
<col width="7%">
<col width="6%">
<col width="7%">
<col width="8%">
<col width="8%">
<col width="8%">
<col width="6%">
<col width="8%">
<col width="6%">
<col width="7%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left">feature_id</th>
<th align="right">timepoint1</th>
<th align="right">timepoint2</th>
<th align="left">resample</th>
<th align="right">N_total_i0</th>
<th align="right">N_total_it</th>
<th align="right">N_light_it</th>
<th align="right">N_heavy_it</th>
<th align="right">EAF</th>
<th align="right">r_net</th>
<th align="right">bi</th>
<th align="right">di</th>
<th align="right">ri</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">taxon_1</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="left">1</td>
<td align="right">1595472105</td>
<td align="right">148586025.4</td>
<td align="right">136810605.5</td>
<td align="right">11775419.83</td>
<td align="right">0.0790913</td>
<td align="right">-1446886080</td>
<td align="right">0.0082567</td>
<td align="right">-0.2456327</td>
<td align="right">-0.2373761</td>
</tr>
<tr class="even">
<td align="left">taxon_2</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="left">1</td>
<td align="right">64576684</td>
<td align="right">10029559.5</td>
<td align="right">9050990.0</td>
<td align="right">978569.49</td>
<td align="right">0.0973733</td>
<td align="right">-54547125</td>
<td align="right">0.0102663</td>
<td align="right">-0.1964979</td>
<td align="right">-0.1862317</td>
</tr>
<tr class="odd">
<td align="left">taxon_3</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="left">1</td>
<td align="right">4488930</td>
<td align="right">461034.7</td>
<td align="right">400320.0</td>
<td align="right">60714.77</td>
<td align="right">0.1314289</td>
<td align="right">-4027895</td>
<td align="right">0.0141209</td>
<td align="right">-0.2417106</td>
<td align="right">-0.2275896</td>
</tr>
<tr class="even">
<td align="left">taxon_4</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="left">1</td>
<td align="right">2494463</td>
<td align="right">379679.5</td>
<td align="right">353589.2</td>
<td align="right">26090.33</td>
<td align="right">0.0685792</td>
<td align="right">-2114784</td>
<td align="right">0.0071192</td>
<td align="right">-0.1953693</td>
<td align="right">-0.1882501</td>
</tr>
<tr class="odd">
<td align="left">taxon_5</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="left">1</td>
<td align="right">9849881</td>
<td align="right">3875688.9</td>
<td align="right">2817314.3</td>
<td align="right">1058374.58</td>
<td align="right">0.2725340</td>
<td align="right">-5974192</td>
<td align="right">0.0318939</td>
<td align="right">-0.1251675</td>
<td align="right">-0.0932736</td>
</tr>
<tr class="even">
<td align="left">taxon_6</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="left">1</td>
<td align="right">697760597</td>
<td align="right">184676166.7</td>
<td align="right">146843360.6</td>
<td align="right">37832806.09</td>
<td align="right">0.2044504</td>
<td align="right">-513084430</td>
<td align="right">0.0229237</td>
<td align="right">-0.1558510</td>
<td align="right">-0.1329272</td>
</tr>
</tbody>
</table>
<p>Some columns contain important, but redundant information. For
example, for each feature <code>timepoint1</code>,
<code>timepoint2</code>, <code>N_total_i0</code>,
<code>N_total_it</code>, <code>r_net</code> and <code>ri</code> are the
same for all rows.</p>
<ul>
<li>
<code>timepoint1</code> and <code>timepoint2</code> are the
timepoints for the growth calculations. For this dataset, we are
comparing day 10 to day 0, so the rates will be in units of “per
day”.</li>
<li>
<code>N_total_i0</code> is the total abundance of each feature at
time 0, and <code>N_total_it</code> is the total abundance of each
feature at time
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>.
<code>r_net</code> is just the copy number difference between the two
time points for each feature, or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>N</mi><mrow><mi>T</mi><mi>O</mi><mi>T</mi><mi>A</mi><mi>L</mi><mi>i</mi><mi>t</mi></mrow></msub><mo>−</mo><msub><mi>N</mi><mrow><mi>T</mi><mi>O</mi><mi>T</mi><mi>A</mi><mi>L</mi><mi>i</mi><mn>0</mn></mrow></msub></mrow><annotation encoding="application/x-tex">N_{TOTALit} - N_{TOTALi0}</annotation></semantics></math>.</li>
<li>
<code>ri</code> is the overall growth rate, where a negative value
indicates more death than birth</li>
</ul>
<p>The remaining columns use the resampled EAF data to determine which
portion of the <code>N_total_it</code> copies correspond to those taking
up the substrate and those that remain unlabeled.</p>
<ul>
<li>
<code>N_light_it</code> comes from equation 3 of Koch, 2018<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://esajournals.onlinelibrary.wiley.com/doi/full/10.1002/ecs2.2090" class="external-link uri"&gt;https://esajournals.onlinelibrary.wiley.com/doi/full/10.1002/ecs2.2090&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a>, and is
the proportion of <code>N_total_it</code> that isn’t labeled.</li>
<li>
<code>N_heavy_it</code> is the proportion of <code>N_total_it</code>
that is labeled, and is roughly
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>N</mi><mrow><mi>T</mi><mi>O</mi><mi>T</mi><mi>A</mi><mi>L</mi><mi>i</mi><mi>t</mi></mrow></msub><mo>*</mo><mi>E</mi><mi>A</mi><mi>F</mi></mrow><annotation encoding="application/x-tex">N_{TOTALit} * EAF</annotation></semantics></math>
</li>
<li>
<code>bi</code> is the per-unit-of-time birth rate, <code>di</code>
is the death rate</li>
</ul>
</div>
<div class="section level3">
<h3 id="summarizing-growth-data">Summarizing Growth Data<a class="anchor" aria-label="anchor" href="#summarizing-growth-data"></a>
</h3>
<p>We can summarize the growth data at a specified
<code>confidence</code> with the <code><a href="../reference/summarize_growth_values.html">summarize_growth_values()</a></code>
function. This function will calculate the mean, sd and confidence
intervals for the birth and death rates, as well as EAF.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarize_growth_values.html">summarize_growth_values</a></span><span class="op">(</span><span class="va">q</span>, confidence <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 351 × 28</span></span></span>
<span><span class="co">#&gt;    feature_id timepoint1 timepoint2  N_total_i0 N_total_it     r_net observed_bi</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> taxon_1             0         10 <span style="text-decoration: underline;">1</span>595<span style="text-decoration: underline;">472</span>105. 148<span style="text-decoration: underline;">586</span>025.   -<span style="color: #BB0000;">1.45</span><span style="color: #949494;">e</span>9     0.010<span style="text-decoration: underline;">9</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> taxon_10            0         10    2<span style="text-decoration: underline;">793</span>486.   1<span style="text-decoration: underline;">058</span>203.   -<span style="color: #BB0000;">1.74</span><span style="color: #949494;">e</span>6     0.025<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> taxon_100           0         10    4<span style="text-decoration: underline;">698</span>016.   1<span style="text-decoration: underline;">933</span>443.   -<span style="color: #BB0000;">2.76</span><span style="color: #949494;">e</span>6     0.047<span style="text-decoration: underline;">6</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> taxon_101           0         10    4<span style="text-decoration: underline;">359</span>459.   1<span style="text-decoration: underline;">617</span>996.   -<span style="color: #BB0000;">2.74</span><span style="color: #949494;">e</span>6     0.064<span style="text-decoration: underline;">2</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> taxon_102           0         10   45<span style="text-decoration: underline;">813</span>796.   6<span style="text-decoration: underline;">260</span>993.   -<span style="color: #BB0000;">3.96</span><span style="color: #949494;">e</span>7     0.004<span style="text-decoration: underline;">02</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> taxon_103           0         10    4<span style="text-decoration: underline;">639</span>329.    <span style="text-decoration: underline;">635</span>123.   -<span style="color: #BB0000;">4.00</span><span style="color: #949494;">e</span>6     0.003<span style="text-decoration: underline;">92</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> taxon_104           0         10   35<span style="text-decoration: underline;">390</span>036.   6<span style="text-decoration: underline;">306</span>709.   -<span style="color: #BB0000;">2.91</span><span style="color: #949494;">e</span>7     0.006<span style="text-decoration: underline;">18</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> taxon_105           0         10  381<span style="text-decoration: underline;">417</span>581.  64<span style="text-decoration: underline;">847</span>518.   -<span style="color: #BB0000;">3.17</span><span style="color: #949494;">e</span>8     0.005<span style="text-decoration: underline;">84</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> taxon_106           0         10    8<span style="text-decoration: underline;">761</span>701.   1<span style="text-decoration: underline;">541</span>086.   -<span style="color: #BB0000;">7.22</span><span style="color: #949494;">e</span>6     0.004<span style="text-decoration: underline;">77</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> taxon_107           0         10    3<span style="text-decoration: underline;">724</span>648.    <span style="text-decoration: underline;">338</span>098.   -<span style="color: #BB0000;">3.39</span><span style="color: #949494;">e</span>6     0.002<span style="text-decoration: underline;">31</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 341 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 21 more variables: observed_di &lt;dbl&gt;, observed_ri &lt;dbl&gt;, successes &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   resampled_N_mean &lt;dbl&gt;, resampled_rnet_mean &lt;dbl&gt;, resampled_bi_mean &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   resampled_bi_sd &lt;dbl&gt;, resampled_bi_lower &lt;dbl&gt;, resampled_bi_upper &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   resampled_di_mean &lt;dbl&gt;, resampled_di_sd &lt;dbl&gt;, resampled_di_lower &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   resampled_di_upper &lt;dbl&gt;, resampled_ri_mean &lt;dbl&gt;, resampled_ri_sd &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   resampled_ri_lower &lt;dbl&gt;, resampled_ri_upper &lt;dbl&gt;, …</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="growth-rate-plots">Growth rate plots<a class="anchor" aria-label="anchor" href="#growth-rate-plots"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_growth_values.html">plot_growth_values</a></span><span class="op">(</span><span class="va">q</span>,</span>
<span>                   confidence <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>                   top <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                   alpha <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>                   error <span class="op">=</span> <span class="st">"ribbon"</span></span>
<span>                   <span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span></code></pre></div>
<p><img src="growth_files/figure-html/unnamed-chunk-15-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="when-growth-cannot-be-calculated">When growth cannot be calculated<a class="anchor" aria-label="anchor" href="#when-growth-cannot-be-calculated"></a>
</h3>
<p>There are a few cases where growth cannot be calculated or the
results can be non-sensical. Some cases result in the entire feature
being unusable, while other cases just remove specific resamples for
that feature while using the remaining features where possible.</p>
<div class="section level4">
<h4 id="no-time-zero-data">No time zero data<a class="anchor" aria-label="anchor" href="#no-time-zero-data"></a>
</h4>
<p>As noted above, taxon_194 has zero abundance at time zero. Therefore,
the growth rate cannot be calculated because any change in abundance
would be considered infinite growth. The intermediate values for these
features can be found in the <code><a href="../reference/get_growth_data.html">get_growth_data()</a></code> function,
but the feature will be omitted entirely from the
<code><a href="../reference/summarize_growth_values.html">summarize_growth_values()</a></code> data.</p>
</div>
<div class="section level4">
<h4 id="negative-eaf-values">Negative EAF values<a class="anchor" aria-label="anchor" href="#negative-eaf-values"></a>
</h4>
<p>This is related to the warning we received above stating there were
31862 resamplings that have “negative EAF values”. While negative EAF
values can be common due to noise, it doesn’t make sense when
calculating
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mrow><mi>L</mi><mi>I</mi><mi>G</mi><mi>H</mi><mi>T</mi><mi>i</mi><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">N_{LIGHTit}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mrow><mi>H</mi><mi>E</mi><mi>A</mi><mi>V</mi><mi>Y</mi><mi>i</mi><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">N_{HEAVYit}</annotation></semantics></math>
values. This happens because
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mrow><mi>L</mi><mi>I</mi><mi>G</mi><mi>H</mi><mi>T</mi><mi>i</mi><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">N_{LIGHTit}</annotation></semantics></math>
gets calculated to actually have more copies than
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mrow><mi>T</mi><mi>O</mi><mi>T</mi><mi>A</mi><mi>L</mi><mi>i</mi><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">N_{TOTALit}</annotation></semantics></math>,
which is impossible, and therefore
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>N</mi><mrow><mi>H</mi><mi>E</mi><mi>A</mi><mi>V</mi><mi>Y</mi><mi>i</mi><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">N_{HEAVYit}</annotation></semantics></math>
will be a negative number of copies, which is also impossible. Below is
from the <code>q@growth$negative_labeled</code> dataframe for taxon_1
explaining the reasoning. Z (equation 4 from Hungate, 2015<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://journals.asm.org/doi/10.1128/aem.02280-15" class="external-link uri"&gt;https://journals.asm.org/doi/10.1128/aem.02280-15&lt;/a&gt;&lt;/p&gt;'><sup>4</sup></a>) is the difference
between the labeled and unlabeled WAD value, so when Z is negative, it
indicates the WAD values were lower for the labeled fractions, likely
due to noise in the SIP process.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q</span><span class="op">@</span><span class="va">growth</span><span class="op">$</span><span class="va">negative_labeled</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">feature_id</span> <span class="op">==</span> <span class="st">"taxon_1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">feature_id</span>, <span class="va">N_total_it</span>, <span class="va">resample</span>, <span class="va">Z</span>, <span class="va">EAF</span>, <span class="va">N_light_it</span>, <span class="va">N_heavy_it</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<caption>First few rows for taxon_1</caption>
<colgroup>
<col width="14%">
<col width="14%">
<col width="11%">
<col width="14%">
<col width="14%">
<col width="14%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th align="left">feature_id</th>
<th align="right">N_total_it</th>
<th align="left">resample</th>
<th align="right">Z</th>
<th align="right">EAF</th>
<th align="right">N_light_it</th>
<th align="right">N_heavy_it</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">taxon_1</td>
<td align="right">148586025</td>
<td align="left">40</td>
<td align="right">-0.0020331</td>
<td align="right">-0.0309272</td>
<td align="right">153190583</td>
<td align="right">-4604557.9</td>
</tr>
<tr class="even">
<td align="left">taxon_1</td>
<td align="right">148586025</td>
<td align="left">56</td>
<td align="right">-0.0023248</td>
<td align="right">-0.0353985</td>
<td align="right">153856294</td>
<td align="right">-5270268.8</td>
</tr>
<tr class="odd">
<td align="left">taxon_1</td>
<td align="right">148586025</td>
<td align="left">80</td>
<td align="right">-0.0001851</td>
<td align="right">-0.0028122</td>
<td align="right">149004712</td>
<td align="right">-418686.2</td>
</tr>
<tr class="even">
<td align="left">taxon_1</td>
<td align="right">148586025</td>
<td align="left">144</td>
<td align="right">-0.0027846</td>
<td align="right">-0.0422962</td>
<td align="right">154883248</td>
<td align="right">-6297222.6</td>
</tr>
<tr class="odd">
<td align="left">taxon_1</td>
<td align="right">148586025</td>
<td align="left">145</td>
<td align="right">-0.0003687</td>
<td align="right">-0.0056138</td>
<td align="right">149421834</td>
<td align="right">-835808.9</td>
</tr>
<tr class="even">
<td align="left">taxon_1</td>
<td align="right">148586025</td>
<td align="left">290</td>
<td align="right">-0.0007288</td>
<td align="right">-0.0110860</td>
<td align="right">150236552</td>
<td align="right">-1650526.9</td>
</tr>
</tbody>
</table>
<p>taxon_1 had a total of 28 resamplings fall into this category, but
the remaining 972 were successful. This number is reflected in the
<code>successes</code> column of
<code><a href="../reference/summarize_growth_values.html">summarize_growth_values()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarize_growth_values.html">summarize_growth_values</a></span><span class="op">(</span><span class="va">q</span>, confidence <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">feature_id</span> <span class="op">==</span> <span class="st">"taxon_1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">feature_id</span>, <span class="va">successes</span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 2</span></span></span>
<span><span class="co">#&gt;   feature_id successes</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> taxon_1          972</span></span></code></pre></div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jeff Kimbrel.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
