<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Standard qSIP EAF Workflow • qSIP2</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Standard qSIP EAF Workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">qSIP2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.20.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/qSIP_workflow.html">Standard qSIP EAF Workflow</a></li>
    <li><a class="dropdown-item" href="../articles/filtering.html">Filtering Features</a></li>
    <li><a class="dropdown-item" href="../articles/resampling.html">Resampling</a></li>
    <li><a class="dropdown-item" href="../articles/EAF.html">EAF Calculations</a></li>
    <li><a class="dropdown-item" href="../articles/growth.html">Growth Workflow</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jeffkimbrel/qSIP2/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Standard qSIP EAF Workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jeffkimbrel/qSIP2/blob/master/vignettes/qSIP_workflow.Rmd" class="external-link"><code>vignettes/qSIP_workflow.Rmd</code></a></small>
      <div class="d-none name"><code>qSIP_workflow.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jeffkimbrel/qSIP2" class="external-link">qSIP2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"qSIP2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] '0.20.3'</span></span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>A complete quantitative stable isotope probing (qSIP) workflow using
the <code>qSIP2</code> package starts with three input files and ends
with calculated excess atom fraction (EAF) values along with a ton of
intermediate data. This vignette will be a high-level walk through of
the major steps with links to more specific vignettes where more detail
is appropriate.</p>
</div>
<div class="section level2">
<h2 id="the-input-files">The Input Files<a class="anchor" aria-label="anchor" href="#the-input-files"></a>
</h2>
<p>Preparing and formatting the input files is often the most tedious
part of any analysis. Our goal with the rigid (and opinionated)
requirements imposed by <code>qSIP2</code> will hopefully streamline the
creation of these files, and automated validation checks can remove many
of the common sources of error or confusion.</p>
<div class="section level3">
<h3 id="source-data">Source Data<a class="anchor" aria-label="anchor" href="#source-data"></a>
</h3>
<p>The source data is the highest level of metadata with a row
corresponding to each original experimental or source material object.
An example source dataframe is included in the <code>qSIP2</code>
package called <code>example_source_df</code>.</p>
<table class="table">
<caption>The first few rows of <code>example_source_df</code>
</caption>
<thead><tr class="header">
<th align="left">source</th>
<th align="right">total_copies_per_g</th>
<th align="right">total_dna</th>
<th align="left">Isotope</th>
<th align="left">Moisture</th>
<th align="left">isotopolog</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">S149</td>
<td align="right">34838665</td>
<td align="right">74.46539</td>
<td align="left">12C</td>
<td align="left">Normal</td>
<td align="left">glucose</td>
</tr>
<tr class="even">
<td align="left">S150</td>
<td align="right">53528072</td>
<td align="right">109.01522</td>
<td align="left">12C</td>
<td align="left">Normal</td>
<td align="left">glucose</td>
</tr>
<tr class="odd">
<td align="left">S151</td>
<td align="right">95774992</td>
<td align="right">182.16852</td>
<td align="left">12C</td>
<td align="left">Normal</td>
<td align="left">glucose</td>
</tr>
<tr class="even">
<td align="left">S152</td>
<td align="right">9126192</td>
<td align="right">23.68963</td>
<td align="left">12C</td>
<td align="left">Normal</td>
<td align="left">glucose</td>
</tr>
<tr class="odd">
<td align="left">S161</td>
<td align="right">41744046</td>
<td align="right">67.62552</td>
<td align="left">12C</td>
<td align="left">Drought</td>
<td align="left">glucose</td>
</tr>
<tr class="even">
<td align="left">S162</td>
<td align="right">49402713</td>
<td align="right">94.21217</td>
<td align="left">12C</td>
<td align="left">Drought</td>
<td align="left">glucose</td>
</tr>
</tbody>
</table>
<p>There are a few required columns for valid source data including a
unique ID (<code>source_mat_id</code>), an <code>isotope</code> and
<code>isotopolog</code> designation for the substrate that had the
label. Additional columns can be added as necessary
(e.g. <code>Moisture</code>) for grouping and filtering later in the
process.</p>
<p>For growth calculations there are three additional requirements -
<code>timepoint</code>, <code>total_abundance</code> and
<code>volume</code>. These are not necessary for the standard EAF
workflow and will instead be addressed in a forthcoming growth
vignette.</p>
<p>Once the dataframe is ready, the next step is to convert it to a
<code>qsip_source_data</code> object. This is one of the main
<code>qSIP2</code> objects to hold and validate the data. Each of the
required columns of metadata is assigned to a column in your dataframe.
For example, below the “Isotope” column in the dataframe is assigned to
the <code>isotope</code> parameter in the <code>qsip_source_data</code>
object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">source_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qsip_source_data.html">qsip_source_data</a></span><span class="op">(</span><span class="va">example_source_df</span>,</span>
<span>  isotope <span class="op">=</span> <span class="st">"Isotope"</span>,</span>
<span>  isotopolog <span class="op">=</span> <span class="st">"isotopolog"</span>,</span>
<span>  source_mat_id <span class="op">=</span> <span class="st">"source"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">source_object</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "qSIP2::qsip_source_data" "S7_object"</span></span></code></pre></div>
<p>This object modifies some of the column names to standard names as
supplied in the above function.</p>
<table class="table">
<thead><tr class="header">
<th align="left">Original Headers</th>
<th align="left">Modified Headers</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">source</td>
<td align="left">source_mat_id</td>
</tr>
<tr class="even">
<td align="left">total_copies_per_g</td>
<td align="left">total_copies_per_g</td>
</tr>
<tr class="odd">
<td align="left">total_dna</td>
<td align="left">total_dna</td>
</tr>
<tr class="even">
<td align="left">Isotope</td>
<td align="left">isotope</td>
</tr>
<tr class="odd">
<td align="left">Moisture</td>
<td align="left">Moisture</td>
</tr>
<tr class="even">
<td align="left">isotopolog</td>
<td align="left">isotopolog</td>
</tr>
</tbody>
</table>
<p>If the column names in the dataframe already match the expected
standard names, then you can skip assigning them and they should be
identified correctly.</p>
<p>A dataframe with the original headers can be recovered using the
<code><a href="../reference/get_dataframe.html">get_dataframe()</a></code> method with the
<code>original_headers = T</code> option.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_dataframe.html">get_dataframe</a></span><span class="op">(</span><span class="va">source_object</span>, original_headers <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Isotope</th>
<th align="left">isotopolog</th>
<th align="left">source</th>
<th align="right">total_copies_per_g</th>
<th align="right">total_dna</th>
<th align="left">Moisture</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">12C</td>
<td align="left">glucose</td>
<td align="left">S149</td>
<td align="right">34838665</td>
<td align="right">74.46539</td>
<td align="left">Normal</td>
</tr>
<tr class="even">
<td align="left">12C</td>
<td align="left">glucose</td>
<td align="left">S150</td>
<td align="right">53528072</td>
<td align="right">109.01522</td>
<td align="left">Normal</td>
</tr>
<tr class="odd">
<td align="left">12C</td>
<td align="left">glucose</td>
<td align="left">S151</td>
<td align="right">95774992</td>
<td align="right">182.16852</td>
<td align="left">Normal</td>
</tr>
<tr class="even">
<td align="left">12C</td>
<td align="left">glucose</td>
<td align="left">S152</td>
<td align="right">9126192</td>
<td align="right">23.68963</td>
<td align="left">Normal</td>
</tr>
<tr class="odd">
<td align="left">12C</td>
<td align="left">glucose</td>
<td align="left">S161</td>
<td align="right">41744046</td>
<td align="right">67.62552</td>
<td align="left">Drought</td>
</tr>
<tr class="even">
<td align="left">12C</td>
<td align="left">glucose</td>
<td align="left">S162</td>
<td align="right">49402713</td>
<td align="right">94.21217</td>
<td align="left">Drought</td>
</tr>
</tbody>
</table>
<p>See <code><a href="../articles/source_data.html">vignette("source_data")</a></code> for more details on working
with source data.</p>
</div>
<div class="section level3">
<h3 id="sample-data">Sample Data<a class="anchor" aria-label="anchor" href="#sample-data"></a>
</h3>
<p>The sample metadata is the next level of detail with one row for each
fraction, or one row for each set of fastq files that were
sequenced.</p>
<table class="table">
<caption>The first few rows of <code>example_sample_df</code>
</caption>
<thead><tr class="header">
<th align="left">sample</th>
<th align="left">source</th>
<th align="right">Fraction</th>
<th align="right">density_g_ml</th>
<th align="right">dna_conc</th>
<th align="right">avg_16S_g_soil</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">149_F1</td>
<td align="left">S149</td>
<td align="right">1</td>
<td align="right">1.778855</td>
<td align="right">0.0000000</td>
<td align="right">4473.7081</td>
</tr>
<tr class="even">
<td align="left">149_F2</td>
<td align="left">S149</td>
<td align="right">2</td>
<td align="right">1.773391</td>
<td align="right">0.0000000</td>
<td align="right">986.6581</td>
</tr>
<tr class="odd">
<td align="left">149_F3</td>
<td align="left">S149</td>
<td align="right">3</td>
<td align="right">1.765742</td>
<td align="right">0.0000000</td>
<td align="right">4002.7026</td>
</tr>
<tr class="even">
<td align="left">149_F4</td>
<td align="left">S149</td>
<td align="right">4</td>
<td align="right">1.759185</td>
<td align="right">0.0000000</td>
<td align="right">3959.7283</td>
</tr>
<tr class="odd">
<td align="left">149_F5</td>
<td align="left">S149</td>
<td align="right">5</td>
<td align="right">1.752629</td>
<td align="right">0.0012413</td>
<td align="right">5725.7319</td>
</tr>
<tr class="even">
<td align="left">149_F6</td>
<td align="left">S149</td>
<td align="right">6</td>
<td align="right">1.746072</td>
<td align="right">0.0128156</td>
<td align="right">7566.2722</td>
</tr>
</tbody>
</table>
<p>Again, there are several necessary columns for valid sample data,
including a unique sample ID (<code>sample_id</code>), the source they
came from (<code>source_mat_id</code>), the fraction ID
(<code>gradient_position</code>), the fraction density
(<code>gradient_pos_density</code>) and a measure of abundance (total
DNA or qPCR copy number) in that fraction
(<code>gradient_pos_amt</code>).</p>
<p>An additional column that can be derived is the percent abundance of
your total sample that is found in each of the fractions. The
<code><a href="../reference/add_gradient_pos_rel_amt.html">add_gradient_pos_rel_amt()</a></code> function can help calculate that
by dividing each fraction abundance by the total abundance for each
source and putting it in a <code>gradient_pos_rel_amt</code> column.</p>
<p>But there is no need to do this if you already have the relative
amounts calculated in your dataframe.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_df</span> <span class="op">&lt;-</span> <span class="va">example_sample_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/add_gradient_pos_rel_amt.html">add_gradient_pos_rel_amt</a></span><span class="op">(</span>source_mat_id <span class="op">=</span> <span class="st">"source"</span>, amt <span class="op">=</span> <span class="st">"avg_16S_g_soil"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Additional <code>gradient_pos_rel_amt</code> column
added</caption>
<colgroup>
<col width="8%">
<col width="8%">
<col width="10%">
<col width="15%">
<col width="12%">
<col width="18%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="left">sample</th>
<th align="left">source</th>
<th align="right">Fraction</th>
<th align="right">density_g_ml</th>
<th align="right">dna_conc</th>
<th align="right">avg_16S_g_soil</th>
<th align="right">gradient_pos_rel_amt</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">149_F1</td>
<td align="left">S149</td>
<td align="right">1</td>
<td align="right">1.778855</td>
<td align="right">0.0000000</td>
<td align="right">4473.7081</td>
<td align="right">0.0001284</td>
</tr>
<tr class="even">
<td align="left">149_F2</td>
<td align="left">S149</td>
<td align="right">2</td>
<td align="right">1.773391</td>
<td align="right">0.0000000</td>
<td align="right">986.6581</td>
<td align="right">0.0000283</td>
</tr>
<tr class="odd">
<td align="left">149_F3</td>
<td align="left">S149</td>
<td align="right">3</td>
<td align="right">1.765742</td>
<td align="right">0.0000000</td>
<td align="right">4002.7026</td>
<td align="right">0.0001149</td>
</tr>
<tr class="even">
<td align="left">149_F4</td>
<td align="left">S149</td>
<td align="right">4</td>
<td align="right">1.759185</td>
<td align="right">0.0000000</td>
<td align="right">3959.7283</td>
<td align="right">0.0001137</td>
</tr>
<tr class="odd">
<td align="left">149_F5</td>
<td align="left">S149</td>
<td align="right">5</td>
<td align="right">1.752629</td>
<td align="right">0.0012413</td>
<td align="right">5725.7319</td>
<td align="right">0.0001643</td>
</tr>
<tr class="even">
<td align="left">149_F6</td>
<td align="left">S149</td>
<td align="right">6</td>
<td align="right">1.746072</td>
<td align="right">0.0128156</td>
<td align="right">7566.2722</td>
<td align="right">0.0002172</td>
</tr>
</tbody>
</table>
<p>Again, we make a <code>qSIP2</code> object for this data, this time
as a <code>qsip_sample_data</code> object. The columns in the dataframe
are assigned to the appropriate parameters, and any column names exactly
matching the parameter name will automatically be identified. Similar to
the <code>source_data</code> above, the names in the
<code>sample_data</code> object will be modified from the original names
to the standardized names.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qsip_sample_data.html">qsip_sample_data</a></span><span class="op">(</span><span class="va">sample_df</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  source_mat_id <span class="op">=</span> <span class="st">"source"</span>,</span>
<span>  gradient_position <span class="op">=</span> <span class="st">"Fraction"</span>,</span>
<span>  gradient_pos_density <span class="op">=</span> <span class="st">"density_g_ml"</span>,</span>
<span>  gradient_pos_amt <span class="op">=</span> <span class="st">"avg_16S_g_soil"</span>,</span>
<span>  gradient_pos_rel_amt <span class="op">=</span> <span class="st">"gradient_pos_rel_amt"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sample_object</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "qSIP2::qsip_sample_data" "S7_object"</span></span></code></pre></div>
<p>See <code><a href="../articles/sample_data.html">vignette("sample_data")</a></code> for more information on
working with sample data including the built-in validations.</p>
</div>
<div class="section level3">
<h3 id="feature-data">Feature Data<a class="anchor" aria-label="anchor" href="#feature-data"></a>
</h3>
<p>Finally, the last of the three necessary input files is a feature
abundance table, aka “OTU table” or “ASV table”. The format of this
dataframe has the unique feature IDs in the first column, and an
additional column for each sample. Each row then contains the whole
number (non-normalized) counts of each feature in each sample.</p>
<p>For now, the validation step defaults to requiring all values be
<code>counts</code> (positive integers), but other <code>type</code>
options include <code>coverage</code> (for working with MAGs or
metagenomes), <code>relative</code> if you already have relative
abundances and <code>normalized</code> if you have spike-ins or another
method that determines the correct abundance in each sample.</p>
<table class="table">
<caption>First bit of <code>example_feature_df</code>
</caption>
<thead><tr class="header">
<th align="left">ASV</th>
<th align="right">149_F1</th>
<th align="right">149_F2</th>
<th align="right">149_F3</th>
<th align="right">149_F4</th>
<th align="right">149_F5</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ASV_1</td>
<td align="right">1245</td>
<td align="right">376</td>
<td align="right">582</td>
<td align="right">1258</td>
<td align="right">692</td>
</tr>
<tr class="even">
<td align="left">ASV_2</td>
<td align="right">1471</td>
<td align="right">569</td>
<td align="right">830</td>
<td align="right">1373</td>
<td align="right">737</td>
</tr>
<tr class="odd">
<td align="left">ASV_3</td>
<td align="right">342</td>
<td align="right">152</td>
<td align="right">211</td>
<td align="right">389</td>
<td align="right">218</td>
</tr>
<tr class="even">
<td align="left">ASV_4</td>
<td align="right">288</td>
<td align="right">119</td>
<td align="right">161</td>
<td align="right">294</td>
<td align="right">157</td>
</tr>
<tr class="odd">
<td align="left">ASV_5</td>
<td align="right">317</td>
<td align="right">108</td>
<td align="right">95</td>
<td align="right">292</td>
<td align="right">164</td>
</tr>
<tr class="even">
<td align="left">ASV_6</td>
<td align="right">201</td>
<td align="right">73</td>
<td align="right">130</td>
<td align="right">250</td>
<td align="right">112</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qsip_feature_data.html">qsip_feature_data</a></span><span class="op">(</span><span class="va">example_feature_df</span>,</span>
<span>  feature_id <span class="op">=</span> <span class="st">"ASV"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">feature_object</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "qSIP2::qsip_feature_data" "S7_object"</span></span></code></pre></div>
<p>See <code><a href="../articles/feature_data.html">vignette("feature_data")</a></code> for more details.</p>
</div>
</div>
<div class="section level2">
<h2 id="the-qsip_data-object">The <code>qsip_data</code> Object<a class="anchor" aria-label="anchor" href="#the-qsip_data-object"></a>
</h2>
<p>The <code>qsip_data</code> class is the main workhorse object in the
<code>qSIP2</code> package. It is built from validated versions of the
three previous objects, and is meant to be a self-contained object with
all of the necessary information for analysis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qsip_data.html">qsip_data</a></span><span class="op">(</span></span>
<span>  source_data <span class="op">=</span> <span class="va">source_object</span>,</span>
<span>  sample_data <span class="op">=</span> <span class="va">sample_object</span>,</span>
<span>  feature_data <span class="op">=</span> <span class="va">feature_object</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">There are 15 source_mat_ids, and they are all shared between the source and sample objects</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">There are 284 sample_ids, and they are all shared between the sample and feature objects</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">qsip_object</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "qSIP2::qsip_data" "S7_object"</span></span></code></pre></div>
<p>This function will report if all <code>source_mat_id</code>s are
shared between the source and sample data, and if all
<code>sample_id</code>s are shared between the sample and feature data.
If it reports there are some unshared ids, you can access them with
<code>get_unshared_ids(qsip_object)</code>, but note that it is just a
warning and does not stop the creation of the qSIP object.</p>
<div class="section level3">
<h3 id="visualizations">Visualizations<a class="anchor" aria-label="anchor" href="#visualizations"></a>
</h3>
<p>Behind the scenes, creation of this object also runs some other
calculations, particularly getting the weighted-average density (WAD)
for each feature in each source, and also the tube relative abundance of
each feature. With these, certain visualizations can be made with
built-in functions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_source_wads.html">plot_source_wads</a></span><span class="op">(</span><span class="va">qsip_object</span>, </span>
<span>                 group <span class="op">=</span> <span class="st">"Moisture"</span>,</span>
<span>                 title <span class="op">=</span> <span class="st">"WAD of each source, grouped by Moisture"</span><span class="op">)</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-15-1.png" width="576"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_sample_curves.html">plot_sample_curves</a></span><span class="op">(</span><span class="va">qsip_object</span>,</span>
<span>                   title <span class="op">=</span> <span class="st">"Normalized density curves for each sample"</span>,</span>
<span>                   facet_by <span class="op">=</span> <span class="st">"isotope"</span>,</span>
<span>                   show_wad <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-16-1.png" width="576"></p>
<p>Another sanity check is making sure the reported density values are
on a reasonably straight line with the gradient position. This can be
plotted, and then a Cook’s distance can be calculated to highlight any
outliers. Note, the ends of the gradient are often flagged as outliers,
although this may not necessarily be the case.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_density_outliers.html">plot_density_outliers</a></span><span class="op">(</span><span class="va">qsip_object</span><span class="op">)</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-17-1.png" width="576"></p>
<div class="alert alert-primary">
<p>The design of the <code>qsip_data</code> object is that it is
contains “slots” for each new analysis step. Although you could create a
new object for each step of the workflow, you can assign the output of
each step back to the original object in order to keep everything
together.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="main-workflow">Main Workflow<a class="anchor" aria-label="anchor" href="#main-workflow"></a>
</h2>
<p>Now that we have a validated <code>qsip_data</code> object, we can
start the main workflow consisting of comparison grouping, filtering,
resampling and finally calculating EAF values.</p>
<div class="section level3">
<h3 id="comparison-grouping">Comparison Grouping<a class="anchor" aria-label="anchor" href="#comparison-grouping"></a>
</h3>
<p>Your <code>qsip_data</code> object likely contains all of your data,
but you may only want to run comparisons on certain subsets. The
<code><a href="../reference/get_comparison_groups.html">get_comparison_groups()</a></code> function attempts to identify and
suggest the sources you may want to compare.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_comparison_groups.html">get_comparison_groups</a></span><span class="op">(</span><span class="va">qsip_object</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"Moisture"</span>,</span>
<span>  isotope <span class="op">=</span> <span class="st">"isotope"</span>,</span>
<span>  source_mat_id <span class="op">=</span> <span class="st">"source_mat_id"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Output of <code><a href="../reference/get_comparison_groups.html">get_comparison_groups()</a></code>
</caption>
<thead><tr class="header">
<th align="left">Moisture</th>
<th align="left">12C</th>
<th align="left">13C</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Normal</td>
<td align="left">S149, S150, S151, S152</td>
<td align="left">S178, S179, S180</td>
</tr>
<tr class="even">
<td align="left">Drought</td>
<td align="left">S161, S162, S163, S164</td>
<td align="left">S200, S201, S202, S203</td>
</tr>
</tbody>
</table>
<p>The <code>group</code> argument here is the most important as it will
define the rows that it thinks constitute a comparison. The
<code>isotope</code> argument is what defines the labeled and unlabeled
values for the comparisons. This can be more complex, particularly if
you have more than one isotopolog, and details will be made available in
the future in another vignette.</p>
<p>The first row shows what “Normal” moisture groups we likely want to
use for unlabeled (S149, S150, S151 and S152) to compare to the labeled
(S178, S179 and S180). Sometimes you may also want to compare the
specific labeled samples in a group to <em>all</em> unlabeled. The
<code>qSIP2</code> package has a convenient way to get those by using
the <code><a href="../reference/get_all_by_isotope.html">get_all_by_isotope()</a></code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_all_by_isotope.html">get_all_by_isotope</a></span><span class="op">(</span><span class="va">qsip_object</span>, <span class="st">"12C"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "S149" "S150" "S151" "S152" "S161" "S162" "S163" "S164"</span></span></code></pre></div>
<div class="alert alert-primary">
<p>The <code><a href="../reference/get_comparison_groups.html">get_comparison_groups()</a></code> function is entirely
informational. There is a more automated way to run the qSIP workflow
using the <code><a href="../reference/run_comparison_groups.html">run_comparison_groups()</a></code> function, and it will be
detailed in another vignette.</p>
</div>
</div>
<div class="section level3">
<h3 id="filter-features">Filter Features<a class="anchor" aria-label="anchor" href="#filter-features"></a>
</h3>
<p>The filter features step does two things. First, it is where the set
of labeled and unlabeled sources are defined for a specific comparison.
Second, it is where you can explicitly say how prevalent a feature must
be to be considered “present” in a source. In other words, you define
the parameters that a feature must be found enough of the replicate
sources, and in enough samples that you can calculate an accurate WAD
value for it.</p>
<p>The <code><a href="../reference/run_feature_filter.html">run_feature_filter()</a></code> function takes a
<code>qsip_data</code> object and these different parameters allowing
you to precisely tailor your filtering results. The more strict the
filtering, the fewer features that will pass the filter.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_feature_filter.html">run_feature_filter</a></span><span class="op">(</span><span class="va">qsip_object</span>,</span>
<span>  unlabeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="../reference/get_all_by_isotope.html">get_all_by_isotope</a></span><span class="op">(</span><span class="va">qsip_object</span>, <span class="st">"12C"</span><span class="op">)</span>,</span>
<span>  labeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S178"</span>, <span class="st">"S179"</span>, <span class="st">"S180"</span><span class="op">)</span>,</span>
<span>  min_unlabeled_sources <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  min_labeled_sources <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min_unlabeled_fractions <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  min_labeled_fractions <span class="op">=</span> <span class="fl">6</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; There are initially 2030 unique feature_ids</span></span>
<span><span class="co">#&gt; 1705 of these have abundance in at least one fraction of one source_mat_id</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; Filtering feature_ids by fraction...</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">1519</span> unlabeled and <span style="color: #BB0000;">1417</span> labeled feature_ids were found in <span style="color: #BB0000;">zero fractions</span> in at least one source_mat_id</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">1440</span> unlabeled and <span style="color: #BB0000;">830</span> labeled feature_ids were found in <span style="color: #BB0000;">too few fractions</span> in at least one source_mat_id</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">299</span> unlabeled and <span style="color: #00BB00;">209</span> labeled feature_ids <span style="color: #00BB00;">passed</span> the fraction filter</span></span>
<span><span class="co">#&gt; In total, <span style="color: #00BB00;">346</span> unique feature_ids <span style="color: #00BB00;">passed</span> the fraction filtering requirements...</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; Filtering feature_ids by source...</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">47</span> unlabeled and <span style="color: #BB0000;">137</span> labeled feature_ids failed the source filter because they were found in <span style="color: #BB0000;">zero sources</span></span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">196</span> unlabeled and <span style="color: #BB0000;">127</span> labeled feature_ids failed the source filter because they were found in <span style="color: #BB0000;">too few sources</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">103</span> unlabeled and <span style="color: #00BB00;">82</span> labeled feature_ids <span style="color: #00BB00;">passed</span> the source filter</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; In total, <span style="color: #00BB00;">64</span> unique feature_ids <span style="color: #00BB00;">passed</span> all fraction and source filtering requirements</span></span></code></pre></div>
<div class="alert alert-primary">
<p>Note, although I said earlier you can overwrite your
<code>qsip_data</code> objects as you go, here it might make sense to
create two versions for the moisture treatments. We’ll take the original
<code>qsip_object</code> and save the filtered Normal dataset to
<code>qsip_normal', and the Drought to</code>qsip_drought`.</p>
</div>
<p>Of the 1,705 features found in the “Normal” data, we can see our
rather strict filtering removed all but 64 features from the dataset. We
can visualize these results on a per-source basis with the
<code><a href="../reference/plot_filter_results.html">plot_filter_results()</a></code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_filter_results.html">plot_filter_results</a></span><span class="op">(</span><span class="va">qsip_normal</span><span class="op">)</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-22-1.png" width="768"></p>
<p>Although a large number of features were removed, we can tell that
the 64 that remained actually still make up a large proportion of the
total abundance in each sample. In <strong>A</strong> above, the
retained features (in blue) make up ~75-85% of the total data, while the
removed data (orange) is the remaining ~15-25%.</p>
<p>In <strong>B</strong>, we can see that a surprisingly large number of
features are found 0 times in many sources (gray) and will therefore
never be present regardless of our filtering choices. And although there
are are ~100-200 features that passed the filtering requirements (blue),
our requirement that <code>min_unlabeled_sources = 6</code> and
<code>min_labeled_sources = 3</code> means that only the features
present in many of the blue slices will be retained, leaving only 64
total.</p>
<p>Let’s do the same comparison with the drought samples.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_drought</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_feature_filter.html">run_feature_filter</a></span><span class="op">(</span><span class="va">qsip_object</span>,</span>
<span>  unlabeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="../reference/get_all_by_isotope.html">get_all_by_isotope</a></span><span class="op">(</span><span class="va">qsip_object</span>, <span class="st">"12C"</span><span class="op">)</span>,</span>
<span>  labeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S200"</span>, <span class="st">"S201"</span>, <span class="st">"S202"</span>, <span class="st">"S203"</span><span class="op">)</span>,</span>
<span>  min_unlabeled_sources <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  min_labeled_sources <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min_unlabeled_fractions <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  min_labeled_fractions <span class="op">=</span> <span class="fl">6</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; There are initially 2030 unique feature_ids</span></span>
<span><span class="co">#&gt; 1877 of these have abundance in at least one fraction of one source_mat_id</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; Filtering feature_ids by fraction...</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">1691</span> unlabeled and <span style="color: #BB0000;">1558</span> labeled feature_ids were found in <span style="color: #BB0000;">zero fractions</span> in at least one source_mat_id</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">1440</span> unlabeled and <span style="color: #BB0000;">1212</span> labeled feature_ids were found in <span style="color: #BB0000;">too few fractions</span> in at least one source_mat_id</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">299</span> unlabeled and <span style="color: #00BB00;">285</span> labeled feature_ids <span style="color: #00BB00;">passed</span> the fraction filter</span></span>
<span><span class="co">#&gt; In total, <span style="color: #00BB00;">367</span> unique feature_ids <span style="color: #00BB00;">passed</span> the fraction filtering requirements...</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; Filtering feature_ids by source...</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">68</span> unlabeled and <span style="color: #BB0000;">82</span> labeled feature_ids failed the source filter because they were found in <span style="color: #BB0000;">zero sources</span></span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">196</span> unlabeled and <span style="color: #BB0000;">171</span> labeled feature_ids failed the source filter because they were found in <span style="color: #BB0000;">too few sources</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">103</span> unlabeled and <span style="color: #00BB00;">114</span> labeled feature_ids <span style="color: #00BB00;">passed</span> the source filter</span></span>
<span><span class="co">#&gt; =+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+</span></span>
<span><span class="co">#&gt; In total, <span style="color: #00BB00;">89</span> unique feature_ids <span style="color: #00BB00;">passed</span> all fraction and source filtering requirements</span></span></code></pre></div>
<p>And only 89 features were retained in the Drought dataset.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_filter_results.html">plot_filter_results</a></span><span class="op">(</span><span class="va">qsip_drought</span><span class="op">)</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-24-1.png" width="768"></p>
<p>Strictly speaking, there is no requirement to do strict filtering,
and it is possible to execute <code><a href="../reference/run_feature_filter.html">run_feature_filter()</a></code> with all
parameters set to <code>1</code>. Filtering was originally recommended
because it dramatically sped up the compute time, but running all vs. a
subset of features in <code>qSIP2</code> has a minimal impact. Further,
setting all values to <code>1</code> and utilizing the
<code>allow_failures</code> flag in the resampling step can provide an
alternative way of deciding which features to keep, rather than just
their prevalence amongst sources/samples. More information is provided
in the Resampling section below, or the
<code><a href="../articles/resampling.html">vignette("resampling")</a></code> vignette.</p>
<p>Importantly, since the relative abundances have already been
calculated for each feature, the subsequent steps in the qSIP pipeline
keep each feature’s computations independent, and the results for a
specific feature will be identical regardless of whether other features
are included or filtered out.</p>
</div>
<div class="section level3">
<h3 id="resampling">Resampling<a class="anchor" aria-label="anchor" href="#resampling"></a>
</h3>
<p>The weighted average density (WAD) values were automatically
calculated during the creation of the <code>qsip_data</code> object
earlier. In order to calculate the confidence interval for the EAF
values, we first need to run a resampling/bootstrapping procedure on the
WAD values. For example, a feature found in 6 unlabeled sources will
have 6 WAD values, and these 6 WAD values are resampled many times to
obtain bootstrapped mean WAD values.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_resampling.html">run_resampling</a></span><span class="op">(</span><span class="va">qsip_normal</span>,</span>
<span>  resamples <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  with_seed <span class="op">=</span> <span class="fl">17</span>,</span>
<span>  progress <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As this step requires some random sampling it is good practice to set
the “seed”. Rather than doing this outside of the function, you can pass
the seed as an argument. If you leave blank, it will generate a random
seed. The seed will generate the same results each time you run the
resampling process.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_normal_17_again</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_resampling.html">run_resampling</a></span><span class="op">(</span><span class="va">qsip_normal</span>,</span>
<span>  resamples <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  with_seed <span class="op">=</span> <span class="fl">17</span>,</span>
<span>  progress <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># two runs are identical</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">qsip_normal</span>, <span class="va">qsip_normal_17_again</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">qsip_normal</span><span class="op">@</span><span class="va">resamples</span><span class="op">$</span><span class="va">l</span><span class="op">[[</span><span class="fl">334</span><span class="op">]</span><span class="op">]</span>, <span class="va">qsip_normal_17_again</span><span class="op">@</span><span class="va">resamples</span><span class="op">$</span><span class="va">l</span><span class="op">[[</span><span class="fl">334</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co"># but individual resamplings within are different</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">qsip_normal</span><span class="op">@</span><span class="va">resamples</span><span class="op">$</span><span class="va">l</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">qsip_normal</span><span class="op">@</span><span class="va">resamples</span><span class="op">$</span><span class="va">l</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>Ressampling of the drought dataset. Notice at this step we are
overwriting the original <code>qsip_drought</code> with the results of
<code><a href="../reference/run_resampling.html">run_resampling()</a></code>, rather than creating new objects.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_drought</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_resampling.html">run_resampling</a></span><span class="op">(</span><span class="va">qsip_drought</span>,</span>
<span>  resamples <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  with_seed <span class="op">=</span> <span class="fl">17</span>,</span>
<span>  progress <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>It is possible to get a resampling error if your filtering is too
strict. If so, consult the <code><a href="../articles/resampling.html">vignette("resampling")</a></code> vignette
and consider running with <code>allow_failures = T</code>.</p>
</div>
<div class="section level3">
<h3 id="eaf-calculations">EAF Calculations<a class="anchor" aria-label="anchor" href="#eaf-calculations"></a>
</h3>
<p>And we are finally at the last main step, calculating and summarizing
the excess atom fraction (EAF) values. There are two functions to run,
the first (<code><a href="../reference/run_EAF_calculations.html">run_EAF_calculations()</a></code>) that calculate EAF for
the observed data and all resamplings, and the second
(<code><a href="../reference/summarize_EAF_values.html">summarize_EAF_values()</a></code>) that summarizes that data at a
chosen confidence interval. These are split into two functions simply
because the <code><a href="../reference/run_EAF_calculations.html">run_EAF_calculations()</a></code> can take a little
longer, allowing different parameters to be tried in
<code><a href="../reference/summarize_EAF_values.html">summarize_EAF_values()</a></code> without having to recalculate
everything.</p>
<p>We’ll also <code><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate()</a></code> to add the original Moisture
condition to each dataframe before we combine them.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_EAF_calculations.html">run_EAF_calculations</a></span><span class="op">(</span><span class="va">qsip_normal</span><span class="op">)</span></span>
<span><span class="va">qsip_drought</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_EAF_calculations.html">run_EAF_calculations</a></span><span class="op">(</span><span class="va">qsip_drought</span><span class="op">)</span></span>
<span></span>
<span><span class="va">normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_EAF_values.html">summarize_EAF_values</a></span><span class="op">(</span><span class="va">qsip_normal</span>, confidence <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Moisture <span class="op">=</span> <span class="st">"Normal"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span>
<span></span>
<span><span class="va">drought</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_EAF_values.html">summarize_EAF_values</a></span><span class="op">(</span><span class="va">qsip_drought</span>, confidence <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Moisture <span class="op">=</span> <span class="st">"Drought"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span>
<span></span>
<span><span class="va">eaf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">normal</span>, <span class="va">drought</span><span class="op">)</span></span></code></pre></div>
<p>We can plot the top 50 by each moisture condition.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_EAF_values.html">plot_EAF_values</a></span><span class="op">(</span><span class="va">qsip_normal</span>, </span>
<span>                top <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                confidence <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>                error <span class="op">=</span> <span class="st">"ribbon"</span>,</span>
<span>                title <span class="op">=</span> <span class="st">"Normal moisture"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-25-1.png" width="480"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_EAF_values.html">plot_EAF_values</a></span><span class="op">(</span><span class="va">qsip_drought</span>, </span>
<span>                top <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                confidence <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>                error <span class="op">=</span> <span class="st">"ribbon"</span>,</span>
<span>                title <span class="op">=</span> <span class="st">"Drought moisture"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-26-1.png" width="480"></p>
<p>The <code><a href="../reference/plot_feature_curves.html">plot_feature_curves()</a></code> function allows us to plot the
tube relative abundances for specific feature IDs. Let’s look at two
with high EAF values, and two with low values.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_feature_curves.html">plot_feature_curves</a></span><span class="op">(</span><span class="va">qsip_normal</span>,</span>
<span>  feature_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ASV_55"</span>, <span class="st">"ASV_84"</span>, <span class="st">"ASV_41"</span>, <span class="st">"ASV_44"</span><span class="op">)</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Density curves of selected features"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-27-1.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="working-with-multiple-qsip-objects-preview">Working with multiple qSIP objects (preview)<a class="anchor" aria-label="anchor" href="#working-with-multiple-qsip-objects-preview"></a>
</h2>
<p>It is possible to work with multiple <code>qsip_data</code> objects
if they are in a list. This will be detailed in a forthcoming vignette,
but here is a sneak peak where we can use the existing
<code><a href="../reference/summarize_EAF_values.html">summarize_EAF_values()</a></code> or <code><a href="../reference/plot_EAF_values.html">plot_EAF_values()</a></code>
functions.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qsip_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Normal"</span> <span class="op">=</span> <span class="va">qsip_normal</span>,</span>
<span>                 <span class="st">"Drought"</span> <span class="op">=</span> <span class="va">qsip_drought</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarize_EAF_values.html">summarize_EAF_values</a></span><span class="op">(</span><span class="va">qsip_list</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Confidence level = 0.9</span></span></code></pre>
<table class="table">
<colgroup>
<col width="5%">
<col width="7%">
<col width="8%">
<col width="13%">
<col width="7%">
<col width="7%">
<col width="12%">
<col width="13%">
<col width="11%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left">group</th>
<th align="left">feature_id</th>
<th align="right">observed_EAF</th>
<th align="right">mean_resampled_EAF</th>
<th align="right">lower</th>
<th align="right">upper</th>
<th align="right">labeled_resamples</th>
<th align="right">unlabeled_resamples</th>
<th align="right">labeled_sources</th>
<th align="right">unlabeled_sources</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Normal</td>
<td align="left">ASV_1</td>
<td align="right">-0.0153107</td>
<td align="right">-0.0157044</td>
<td align="right">-0.0516543</td>
<td align="right">0.0236518</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">3</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">Drought</td>
<td align="left">ASV_1</td>
<td align="right">-0.0333856</td>
<td align="right">-0.0330890</td>
<td align="right">-0.0808509</td>
<td align="right">0.0161212</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">4</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">Normal</td>
<td align="left">ASV_10</td>
<td align="right">0.1126260</td>
<td align="right">0.1121874</td>
<td align="right">0.0848992</td>
<td align="right">0.1400368</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">3</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">Drought</td>
<td align="left">ASV_10</td>
<td align="right">0.0543136</td>
<td align="right">0.0541364</td>
<td align="right">0.0303215</td>
<td align="right">0.0776436</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">4</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">Drought</td>
<td align="left">ASV_100</td>
<td align="right">-0.0892684</td>
<td align="right">-0.0895131</td>
<td align="right">-0.1488307</td>
<td align="right">-0.0349891</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">3</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">Drought</td>
<td align="left">ASV_102</td>
<td align="right">-0.0407091</td>
<td align="right">-0.0407609</td>
<td align="right">-0.0907747</td>
<td align="right">0.0088904</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">4</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">Normal</td>
<td align="left">ASV_11</td>
<td align="right">0.3749260</td>
<td align="right">0.3743849</td>
<td align="right">0.3392976</td>
<td align="right">0.4094196</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">3</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">Drought</td>
<td align="left">ASV_11</td>
<td align="right">0.2629470</td>
<td align="right">0.2622983</td>
<td align="right">0.2041474</td>
<td align="right">0.3099201</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">4</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">Normal</td>
<td align="left">ASV_114</td>
<td align="right">0.1926455</td>
<td align="right">0.1918100</td>
<td align="right">0.1234247</td>
<td align="right">0.2683575</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">3</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">Drought</td>
<td align="left">ASV_114</td>
<td align="right">0.2160639</td>
<td align="right">0.2167657</td>
<td align="right">0.1304984</td>
<td align="right">0.2998898</td>
<td align="right">1000</td>
<td align="right">1000</td>
<td align="right">3</td>
<td align="right">7</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_EAF_values.html">plot_EAF_values</a></span><span class="op">(</span><span class="va">qsip_list</span>, </span>
<span>                top <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                confidence <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>                error <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Confidence level = 0.9</span></span></code></pre></div>
<p><img src="qSIP_workflow_files/figure-html/unnamed-chunk-31-1.png" width="768"></p>
</div>
<div class="section level2">
<h2 id="piped-workflow">Piped Workflow<a class="anchor" aria-label="anchor" href="#piped-workflow"></a>
</h2>
<p>Although the workflow is often easier to understand and troubleshoot
when broken into individual steps, it is possible to run the entire
workflow in a single pipe.</p>
<p>For example, the normal moisture data could be filtered, resampled
and have EAF values calculated in a single pipe.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># example code, not executed. Default values are not shown.</span></span>
<span></span>
<span><span class="va">qsip_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_feature_filter.html">run_feature_filter</a></span><span class="op">(</span><span class="va">qsip_object</span>,</span>
<span>  unlabeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="../reference/get_all_by_isotope.html">get_all_by_isotope</a></span><span class="op">(</span><span class="va">qsip_object</span>, <span class="st">"12C"</span><span class="op">)</span>,</span>
<span>  labeled_source_mat_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"S178"</span>, <span class="st">"S179"</span>, <span class="st">"S180"</span><span class="op">)</span>,</span>
<span>  min_unlabeled_sources <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  min_labeled_sources <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/run_resampling.html">run_resampling</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/run_EAF_calculations.html">run_EAF_calculations</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jeff Kimbrel.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
