<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="qSIP2">
<title>Adding and validating sample data • qSIP2</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Adding and validating sample data">
<meta property="og:description" content="qSIP2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">qSIP2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.4.9005</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/sample_data.html">Adding and validating sample data</a>
    <a class="dropdown-item" href="../articles/abundance_data.html">Working with and validating abundance data</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/index.html">More articles...</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jeffkimbrel/qSIP2/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Adding and validating sample data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jeffkimbrel/qSIP2/blob/HEAD/vignettes/sample_data.Rmd" class="external-link"><code>vignettes/sample_data.Rmd</code></a></small>
      <div class="d-none name"><code>sample_data.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>In <code>qSIP2</code>, “sample data” refers to any high level
metadata associated with either an experiment or the individual
fractions. This vignette will show available tools to format and
validate your sample data for the <code>qsip_sample_data</code> class in
the <code>qSIP2</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jeffkimbrel/qSIP2" class="external-link">qSIP2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="what-is-a-sample">What is a sample?<a class="anchor" aria-label="anchor" href="#what-is-a-sample"></a>
</h3>
<p>The word <strong>sample</strong> typically refers to the biological
or environmental entity the DNA was isolated from as well as the single
sequencing run tied to that <strong>sample</strong>. In qSIP, however,
because there are multiple sequencing runs per biological subject, the
term <strong>sample</strong> has historically been reserved for
sequencing of each fraction. In practice, this means you will have many
<strong>samples</strong> for each single biological subject.</p>
<p>Traditionally, the original biological subject in qSIP has been
simply called a <strong>replicate</strong> or “rep”. In the proposed
MISIP<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://www.biorxiv.org/content/10.1101/2023.07.13.548835v1" class="external-link uri"&gt;https://www.biorxiv.org/content/10.1101/2023.07.13.548835v1&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a>
standards, the concept of a <strong>rep</strong> has been replaced with
<code>source_mat_id</code>.</p>
<p>For example, if you were identifying active community members from a
pot of soil in triplicate, you may have 3 soil pot <strong>reps</strong>
that were given <sup>16</sup>O (maybe named “A_16O”, “B_16O” and
“C_16O”), and then 3 <strong>reps</strong> with <sup>18</sup>O (maybe
named “A_18O”, “B_18O” and “C_18O”). In this experiment, you would have
6 totals <strong>reps</strong>. If each <strong>rep</strong> was
separated into 20 fractions and sequenced, then you would have 120
(6x20) <strong>samples</strong>.</p>
<p>If working with amplicon data, each fastq file would be a
<strong>sample</strong>, or each column in an ASV/OTU table (when taxa
are the rows).</p>
</div>
</div>
<div class="section level2">
<h2 id="creating-a-standardized-sample-data-tibble">Creating a standardized sample data tibble<a class="anchor" aria-label="anchor" href="#creating-a-standardized-sample-data-tibble"></a>
</h2>
<p>To standardize the qSIP workflow, column names should adhere as
closely to MISIP standards where possible. Absolute requirements are
slightly different than the official MISIP standards, and for the
<code>qSIP2</code> package would be the <code>source_mat_id</code>,
<code>isotope</code>, <code>gradient_position</code>,
<code>gradient_pos_density</code> and <code>gradient_pos_rel_amt</code>
fields.</p>
<div class="section level3">
<h3 id="isotope-and-isotopolog_label-fields">isotope and isotopolog_label fields<a class="anchor" aria-label="anchor" href="#isotope-and-isotopolog_label-fields"></a>
</h3>
<p>In traditional qSIP the <code>isotope</code> field has been populated
with either the light (e.g. 16O) or heavy (e.g. 18O) isotope depending
on the substrate used in that rep or <code>source_mat_id</code>. In
MISIP standards, only the heavy isotope is listed under the
<code>isotope</code> field, and then a secondary field
<code>isotopolog_label</code> is used to designate whether the replicate
used a substrate with “natural abundance” (i.e. “light”) or
“isotopically labeled” (i.e. “heavy”) isotopes.</p>
<p>In the <code>qSIP2</code> package, either method can be used. If the
<code>isotopolog_label</code> is missing from your dataset then it will
assume both the light and heavy isotopes are present in the
<code>isotope</code> field. But, if you do have an
<code>isotopolog_label</code> field, then only the heavy isotope
designation is allowed in the <code>isotope</code> field and the
dataframe will not pass validation checks if there are light isotopes
listed there.</p>
<p>Conversion between these two objects can be done with the
<code><a href="../reference/add_isotopolog_label.html">add_isotopolog_label()</a></code> or
<code><a href="../reference/remove_isotopolog_label.html">remove_isotopolog_label()</a></code> functions.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># original data has a mixture of isotopes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sample_data_nonMISIP</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   rep   isotope fraction density dna_conc</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1_16O 16O            7    1.70    0.775</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1_16O 16O            8    1.69    1.89 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 1_16O 16O            9    1.69    1.34 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 1_16O 16O           10    1.68    0.641</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 1_16O 16O           11    1.68    0.372</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 1_16O 16O           12    1.67    0.282</span></span>
<span></span>
<span><span class="va">sample_data_nonMISIP</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">isotope</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   isotope     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 16O        38</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 18O        34</span></span>
<span></span>
<span><span class="co"># new data has only one isotope and a mixture of isotopolog_label </span></span>
<span><span class="va">df_with_labels</span> <span class="op">=</span> <span class="fu"><a href="../reference/add_isotopolog_label.html">add_isotopolog_label</a></span><span class="op">(</span><span class="va">sample_data_nonMISIP</span>, </span>
<span>                                      isotope <span class="op">=</span> <span class="st">"isotope"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">df_with_labels</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">isotope</span>, <span class="va">isotopolog_label</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   isotope isotopolog_label         n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 18O     isotopically labeled    34</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 18O     natural abundance       38</span></span>
<span></span>
<span><span class="co"># after converting back the isotope column is a mixture again</span></span>
<span><span class="fu"><a href="../reference/remove_isotopolog_label.html">remove_isotopolog_label</a></span><span class="op">(</span><span class="va">df_with_labels</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">isotope</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   isotope     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 16O        38</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 18O        34</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gradient_pos_rel_amt-field">gradient_pos_rel_amt field<a class="anchor" aria-label="anchor" href="#gradient_pos_rel_amt-field"></a>
</h3>
<p>A requirement for qSIP is the <code>gradient_pos_rel_amt</code>
field, which gives the percent amount that a fraction has of the whole.
The preferred method is given in qPCR copy numbers, but DNA
concentrations can be used as well.</p>
<p>For example, if there are 100,000 total 16S copies in a replicate as
determined by qPCR, and 15,000 copies in fraction 7, then the
<code>gradient_pos_rel_amt</code> value for fraction 7 would be 0.15
(15,000 / 100,000). Similarly, if you had 25ng total DNA used for
density separation, and fraction 7 had 3.75 ng DNA recovered, then
<code>gradient_pos_rel_amt</code> would also be 0.15 (3.75 / 25).</p>
<p>Ideally, all of the <code>gradient_pos_rel_amt</code> for a given
<code>source_mat_id</code> should add up to 1, but there are situations
where it might be less than 1. For example, if you removed some
fractions because they didn’t sequence well or there was some other
reason to remove a fraction. If you have 16S or DNA concentrations for
these removed samples they would be subtracted from the total. So, for
our example above, if fraction 7 needed to be removed, then the total
for all fractions of that <code>source_mat_id</code> would only be 0.85.
Another situation would be that although you added 25ng to a
centrifugation, you only recovered 20ng and the rest was lost<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;&lt;a href=&quot;https://www.visitlex.com/guides/post/common-geeky-bourbon-terms-explained/#:~:text=Angel's%20Share&quot; class=&quot;external-link&quot;&gt;The
angel’s share and devil’s cut&lt;/a&gt;&lt;/p&gt;"><sup>2</sup></a>. The total
fractions within a <code>source_mat_id</code> within should never be
greater than 1.</p>
<p>If your sample data tibble does not have the
<code>gradient_pos_rel_amt</code> there is a function that can add it
for you.</p>
<p>The <code><a href="../reference/add_gradient_pos_rel_amt.html">add_gradient_pos_rel_amt()</a></code> can create this column
from either a column of qPCR totals, or DNA concentrations. Because this
function can not know if there are missing fractions, the totals per
<code>source_mat_id</code> will be equal to 1. If you do have fractions
that you want removed, keep them in the dataframe for this
<code><a href="../reference/add_gradient_pos_rel_amt.html">add_gradient_pos_rel_amt()</a></code> and then remove them after.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># gradient_pos_rel_amt to a dataframe without this column </span></span>
<span><span class="fu"><a href="../reference/add_gradient_pos_rel_amt.html">add_gradient_pos_rel_amt</a></span><span class="op">(</span><span class="va">sample_data_nonMISIP</span>,</span>
<span>                         source_mat_id <span class="op">=</span> <span class="st">"rep"</span>, <span class="co"># the source_mat_id or per-tube grouping</span></span>
<span>                         amt <span class="op">=</span> <span class="st">"dna_conc"</span> <span class="co"># the column with total amounts</span></span>
<span>                         <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   rep   isotope fraction density dna_conc gradient_pos_rel_amt</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1_16O 16O            7    1.70    0.775               0.146 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1_16O 16O            8    1.69    1.89                0.357 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 1_16O 16O            9    1.69    1.34                0.253 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 1_16O 16O           10    1.68    0.641               0.121 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 1_16O 16O           11    1.68    0.372               0.070<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 1_16O 16O           12    1.67    0.282               0.053<span style="text-decoration: underline;">2</span></span></span>
<span></span>
<span><span class="co"># adding to a dataframe with an existing gradient_pos_rel_amt will give an error</span></span>
<span><span class="fu"><a href="../reference/add_gradient_pos_rel_amt.html">add_gradient_pos_rel_amt</a></span><span class="op">(</span><span class="va">sample_data_MISIP</span>,</span>
<span>                         amt <span class="op">=</span> <span class="st">"gradient_pos_dna"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in add_gradient_pos_rel_amt(sample_data_MISIP, amt = "gradient_pos_dna"): <span style="color: #BB0000;">ERROR: gradient_pos_rel_amt already exists! Set overwrite = TRUE if you want to overwrite</span></span></span>
<span></span>
<span><span class="co"># adding overwrite = T removes the error</span></span>
<span><span class="fu"><a href="../reference/add_gradient_pos_rel_amt.html">add_gradient_pos_rel_amt</a></span><span class="op">(</span><span class="va">sample_data_MISIP</span>,</span>
<span>                         amt <span class="op">=</span> <span class="st">"gradient_pos_dna"</span>,</span>
<span>                         overwrite <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; WARNING: gradient_pos_rel_amt already exists and will be overwritten</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   source_mat_id isotope isotopolog_label  gradient_position gradient_pos_density</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1_16O         18O     natural abundance                 7                 1.70</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1_16O         18O     natural abundance                 8                 1.69</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 1_16O         18O     natural abundance                 9                 1.69</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 1_16O         18O     natural abundance                10                 1.68</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 1_16O         18O     natural abundance                11                 1.68</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 1_16O         18O     natural abundance                12                 1.67</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: gradient_pos_dna &lt;dbl&gt;, gradient_pos_rel_amt &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="make-a-qsip-sample-data-object">Make a qSIP sample data object<a class="anchor" aria-label="anchor" href="#make-a-qsip-sample-data-object"></a>
</h2>
<p>The <code>qsip_sample_data</code> constructor builds a sample data
object from a dataframe. The <code>qSIP2</code> package comes with a
basic example tibble called <code>sample_data_MISIP</code>, and some of
the columns are already formatted with proper MISIP headers.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#qdf = qsip_sample_data(sample_data_MISIP)</span></span></code></pre></div>
<p>And that’s that, for now.</p>
<div class="section level3">
<h3 id="isotope-values">Isotope Values<a class="anchor" aria-label="anchor" href="#isotope-values"></a>
</h3>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jeff Kimbrel.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
